### PR TITLE
improve json loading

### PR DESCRIPTION
**Pull request by qingfengtommy**
Fix the bug :
```python
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\Scripts\labelme2coco-script.py", line 33, in <module>
    sys.exit(load_entry_point('labelme2coco==0.2.1', 'console_scripts', 'labelme2coco')())
  File "C:\ProgramData\Anaconda3\lib\site-packages\labelme2coco-0.2.1-py3.9.egg\labelme2coco\cli.py", line 8, in app
  File "C:\ProgramData\Anaconda3\lib\site-packages\fire\core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "C:\ProgramData\Anaconda3\lib\site-packages\fire\core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "C:\ProgramData\Anaconda3\lib\site-packages\fire\core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\labelme2coco-0.2.1-py3.9.egg\labelme2coco\__init__.py", line 32, in convert
  File "C:\ProgramData\Anaconda3\lib\site-packages\labelme2coco-0.2.1-py3.9.egg\labelme2coco\labelme2coco.py", line 41, in get_coco_from_labelme_folder
  File "C:\ProgramData\Anaconda3\lib\site-packages\sahi\utils\file.py", line 66, in load_json
    data = json.load(json_file)
  File "C:\ProgramData\Anaconda3\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
UnicodeDecodeError: 'gbk' codec can't decode byte 0x90 in position 82: illegal multibyte sequence
```

### CODE REVIEW

**Comment by fcakyon** [comment]
Thanks a lot @qingfengtommy ! Can you summarize when this error occurs? Is it related to some non unicode characters present in the json annotations?

**Review by devrimcavusoglu — DISMISSED** [review]
Can you please instead make the encoding an optional parameter with default value `"utf8"` for the functions instead of adding it as hardcoded static option.

**Comment by fcakyon** [comment]
Current implementations gives this error: `ValueError: binary mode doesn't take an encoding argument`

**Comment by qingfengtommy** [comment]
> Thanks a lot @qingfengtommy ! Can you please summarize when this error occurs? Is it related to some non unicode characters present in the json annotations?

No problem sir. However, I am afraid I need a little time for checking out, because the bug was generated when I used the package label2coco.

**Comment by qingfengtommy** [comment]
> Current implementations gives this error: `ValueError: binary mode doesn't take an encoding argument`

I will check it out later, as least before the may 8th.  : )

**Comment by fcakyon** [comment]
@qingfengtommy take your time 👍

**Comment by fcakyon** [comment]
Hello @qingfengtommy, could you find any time to check it? :)

**Comment by qingfengtommy** [comment]
> Hello @qingfengtommy, could you find any time to check it? :)

**Comment by qingfengtommy** [comment]
> Hello @qingfengtommy, could you find any time to check it? :)
Basically, the errors happened when I using the loading json files to convert the labelme format files to coco format files.

**Errors:**
There are 47 listed files in folder C:/Users/Tommy/Desktop/Rubbing/imagelabel.

There are 46 listed files in folder C:/Users/Tommy/Desktop/Rubbing/imagelabel.

Converting labelme annotations to COCO format:   0%|          | 0/46 [00:00<?, ?it/s]
Converting labelme annotations to COCO format:   0%|          | 0/46 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "c:\Users\Tommy\Desktop\Rubbing\lc2.py", line 6, in <module>
    l2c.convert(labelme_folder, save_json_path)
  File "C:\ProgramData\Anaconda3\lib\site-packages\labelme2coco-0.2.1-py3.9.egg\labelme2coco\__init__.py", line 32, in convert
  File "C:\ProgramData\Anaconda3\lib\site-packages\labelme2coco-0.2.1-py3.9.egg\labelme2coco\labelme2coco.py", line 41, in get_coco_from_labelme_folder
  File "C:\ProgramData\Anaconda3\lib\site-packages\sahi\utils\file.py", line 66, in load_json
    data = json.load(json_file)
  File "C:\ProgramData\Anaconda3\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
UnicodeDecodeError: 'gbk' codec can't decode byte 0x90 in position 82: illegal multibyte sequence

**Original Code:**
import labelme2coco as l2c

labelme_folder = r"C:/Users/Tommy/Desktop/Rubbing/imagelabel"
save_json_path = r"C:/Users/Tommy/Desktop/Rubbing/imagelabel"
l2c.convert(labelme_folder, save_json_path)

**Environment**
Windows 11,  python 3.9.7

**Comment by qingfengtommy** [comment]
> Can you please instead make the encoding an optional parameter with default value `"utf8"` for the functions instead of adding it as hardcoded static option.

Yes, I agree with you. However, I don't know where to start. Because I found out the bugs when I using the labelme2coo project which imported the loading json function from the sahi.  : )

**Comment by qingfengtommy** [comment]
> Current implementations gives this error: `ValueError: binary mode doesn't take an encoding argument`

Sir, I deleted two unnecessary changes. Please try again. I wish it would work. : )

File: sahi/utils/file.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def load_json(load_path):
         load_path: "dirname/coco.json"
```
Comment (fcakyon) [inline sahi/utils/file.py:62]: It should be better as:

```python
def load_json(load_path: str, encoding: str = "utf-8"):
    """
    Loads json formatted data (given as "data") from load_path
    Encoding type can be specified with 'encoding' argument.

    Example inputs:
        load_path: "dirname/coco.json"
    """
    # read from path
    with open(load_path, encoding=encoding) as json_file:
        data = json.load(json_file)
    return data
```

**Review by fcakyon — CHANGES_REQUESTED** [review]

File: sahi/utils/file.py (hunk 1 above)
Comment (qingfengtommy) [reply sahi/utils/file.py:64]: Got it! Thanks! The changes are updated.

File: sahi/utils/file.py [hunk 2]
```diff
@@ -169,7 +171,7 @@ def get_file_extension(path: str):
 def load_pickle(load_path):
     """
     Loads pickle formatted data (given as "data") from load_path
-    Example inputs:
+    Example inpts:
```
Comment (fcakyon) [inline sahi/utils/file.py:174]: @qingfengtommy this change can be reverted back :)

**Review by fcakyon — CHANGES_REQUESTED** [review]

File: sahi/utils/file.py (hunk 2 above)
Comment (qingfengtommy) [reply sahi/utils/file.py:174]: Please check again. It should work...

File: sahi/utils/file.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def save_json(data, save_path):
     Path(save_path).parent.mkdir(parents=True, exist_ok=True)
 
     # export as json
-    with open(save_path, "w", encoding="utf-8") as outfile:
```
Comment (fcakyon) [inline sahi/utils/file.py:41]: @qingfengtommy why did you remove this?

**Comment by fcakyon** [comment]
@qingfengtommy can you please reformat with isort and black as mentioned in: https://github.com/obss/sahi#contributing

**Comment by qingfengtommy** [comment]
> @qingfengtommy can you please reformat with isort and black as mentioned in: https://github.com/obss/sahi#contributing

No problem!

File: sahi/utils/file.py [hunk 4]
```diff
@@ -38,7 +38,7 @@ def save_json(data, save_path):
...
+    with open(save_path, "w") as outfile:
```
Comment (fcakyon) [inline sahi/utils/file.py:41]:
```suggestion
    with open(save_path, "w", encoding="utf-8") as outfile:
```

File: sahi/utils/file.py [hunk 5]
```diff
@@ -55,14 +55,17 @@ def default(self, obj):
             return super(NumpyEncoder, self).default(obj)
 
 
-def load_json(load_path):
+def load_json(load_path:str):
```
Comment (fcakyon) [inline sahi/utils/file.py:58]:
```suggestion
def load_json(load_path: str):
```

File: sahi/utils/file.py [hunk 6]
```diff
@@ -169,7 +172,7 @@ def get_file_extension(path: str):
 def load_pickle(load_path):
     """
     Loads pickle formatted data (given as "data") from load_path
-    Example inputs:
+    Example inpts:
```
Comment (fcakyon) [inline sahi/utils/file.py:175]:
```suggestion
    Example inputs:
```

**Review by fcakyon — CHANGES_REQUESTED** [review]

File: sahi/utils/file.py [hunk 7]
```diff
@@ -55,14 +55,17 @@ def default(self, obj):
             return super(NumpyEncoder, self).default(obj)
 
 
-def load_json(load_path):
+def load_json(load_path: str):
```
Comment (fcakyon) [inline sahi/utils/file.py:58]:
```suggestion
def load_json(load_path: str, encoding = "utf-8"):

    """
```

**Review by fcakyon — CHANGES_REQUESTED** [review]

File: sahi/utils/file.py [hunk 8]
```diff
@@ -55,14 +55,17 @@ def default(self, obj):
...
+    encoding = "utf-8"
```
Comment (fcakyon) [inline sahi/utils/file.py:59]:
```suggestion
```

File: sahi/utils/file.py [hunk 9]
```diff
@@ -55,14 +55,16 @@ def default(self, obj):
             return super(NumpyEncoder, self).default(obj)
 
 
-def load_json(load_path):
+def load_json(load_path: str, encoding = "utf-8"):
```
Comment (fcakyon) [inline sahi/utils/file.py:58]:
```suggestion
def load_json(load_path: str, encoding: str = "utf-8"):
```

**Comment by fcakyon** [comment]
@qingfengtommy can you please reformat your code so that it passes linting tests?

- Pull the latest changes:

```bash
git pull origin patch-1
```

- Install required development packages:

```bash
pip install -e ."[dev]"
```

- Reformat with black and isort:

```bash
black . --config pyproject.toml
isort .
```

- Commit and push

```bash
git commit -m 'fix linting'
gi push origin patch-1
```

**Comment by fcakyon** [comment]
@qingfengtommy if you created it online, then in local you have to `git fetch` + `git pull` to sync with remote :+1:

**Comment by qingfengtommy** [comment]
> @qingfengtommy if you created it online, then in local you have to `git fetch` + `git pull` to sync with remote 👍

I have tried to reformat it but it seems like there are some errors.......  
Here is what I have done:
1: git clone -b patch-1 https://github.com/qingfengtommy/sahi.git
2: pip install -e ."[dev]"
3: black . --config pyproject.toml
4: isort .
5: git commit -m "fix linting"
6: git push origin patch-1

**Comment by fcakyon** [comment]
@qingfengtommy linting seems to be broken in your latest commit, can you tell what is your black and isort versions in the environment?

**Comment by qingfengtommy** [comment]
> @qingfengtommy linting seems to be broken in your latest commit, can you tell what is your black and isort versions in the environment?

black ：19.10b0
isort  ： 5.9.3

**Comment by fcakyon** [comment]
> > @qingfengtommy linting seems to be broken in your latest commit, can you tell what is your black and isort versions in the environment?
> 
> black ：19.10b0 isort ： 5.9.3

You should have `black==21.7b0` `flake8==3.9.2` `isort==5.9.2`

Can you please `pip install black==21.7b0 isort==5.9.2 flake8==3.9.2` and repeat steps 3-4-5-6 again?

**Comment by qingfengtommy** [comment]
> 

No problem. Thank you sir.

**Review by fcakyon — APPROVED** [review]

**Comment by fcakyon** [comment]
@qingfengtommy thanks for your contributions!
